modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C

```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
-  pminusa[0] = eigvals[0] - lam1;
-  pminusa[1] = eigvals[1] - lam1;
-  pminusa[2] = eigvals[2] - lam2;
-  // p = a + alpha*n3 + beta*n5, and for Kuhn-Tucker we need alpha and beta non-negative
-  if (dot(pminusa, n1xn5)/dot(n[3], n1xn5) >= 0 && dot(pminusa, n1xn3)/dot(n[5], n1xn3) >= 0)
+  if (trial == 5)
   {
-    returned_stress(0, 0) = returned_stress(1, 1) = lam1;
-    returned_stress(2, 2) = lam2;
-    returned_stress = eigvecs*returned_stress*(eigvecs.transpose());
-    act[3] = act[5] = true;
-    return;
+    sinphi = std::sin(phi(intnl_old));
+    cosphi = std::cos(phi(intnl_old));
+    coh = cohesion(intnl_old);
+    cohcos = coh*cosphi;
+    yieldFunctionEigvals(eigvals[0], eigvals[1], eigvals[2], sinphi, cohcos, yf);
+    Moose::err << "Trial stress = \n";
+    trial_stress.print(Moose::err);
+    Moose::err << "which has eigenvalues = " << eigvals[0] << " " << eigvals[1] << " " << eigvals[2] << "\n";
+    Moose::err << "and yield functions = " << yf[0] << " " << yf[1] << " " << yf[2] << " " << yf[3] << " " << yf[4] << " " << yf[5] << "\n";
+    Moose::err << "Internal parameter = " << intnl_old << "\n";
+    mooseError("TensorMechanicsPlasticMohrCoulombMulti: FAILURE!  You probably need to implement a line search if your hardening is too severe, or you need to tune your tolerances (eg, yield_function_tolerance should be a little smaller than (young modulus)*ep_plastic_tolerance).\n");
+    return false;
   }
 
-  // 000100 (plane)
-  // the returned point, "a" is defined by f3=0 and
-  // a = p - alpha*n3
-  Real alpha = 0.5*eigvals[2]*(1 + sinphi) + 0.5*eigvals[0]*(-1 + sinphi) - coh*cosphi;
-  alpha /= 0.5*n[3][2]*(1 + sinphi) + 0.5*n[3][0]*(-1 + sinphi);
-  returned_stress(0, 0) = eigvals[0] - alpha*n[3][0];
-  returned_stress(1, 1) = eigvals[1] - alpha*n[3][1];
-  returned_stress(2, 2) = eigvals[2] - alpha*n[3][2];
+  // success
+
+  returned_intnl = intnl_old;
+  for (unsigned i = 0 ; i < 6 ; ++i)
+    returned_intnl += dpm[i];
+  for (unsigned i = 0 ; i < 6 ; ++i)
+      for (unsigned j = 0 ; j < 3 ; ++j)
+        delta_dp(j, j) += dpm[i]*norm[i][j];
   returned_stress = eigvecs*returned_stress*(eigvecs.transpose());
-  act[3] = true;
-  return;
+  delta_dp = eigvecs*delta_dp*(eigvecs.transpose());
+  return true;
 }
 
-Real
-TensorMechanicsPlasticMohrCoulombMulti::dot(const std::vector<Real> & a, const std::vector<Real> & b) const
-{
-  return a[0]*b[0] + a[1]*b[1] + a[2]*b[2];
-}
 
-void
-TensorMechanicsPlasticMohrCoulombMulti::cross(const std::vector<Real> & a, const std::vector<Real> & b, std::vector<Real> & c) const
+
+bool
+TensorMechanicsPlasticMohrCoulombMulti::returnTip(const std::vector<Real> & eigvals, const std::vector<std::vector<Real> > & n, std::vector<Real> & dpm, RankTwoTensor & returned_stress, const Real & intnl_old, Real & sinphi, Real & cohcos, const Real & initial_guess, bool & nr_converged, const Real & ep_plastic_tolerance, std::vector<Real> & yf) const
```

Comment:
Is there a reason you are not using `RealVectorValue` here? It would seem like a more natural choice for a vector that is representing spatial components (i.e. size == 3, combined with a LIBMESH_DIM == 3 check)